Can't install ccls server

**Describe the bug**
Can't install ccls server

**Expected behavior**
Command line execution ":LspInstall ccls"

**Screenshots**
![image](https://user-images.githubusercontent.com/44520002/146349133-7d5c9d7a-27bc-4d55-a4df-1c6d2603ea7a.png)
![image](https://user-images.githubusercontent.com/44520002/146349401-412147fd-7e45-417f-8a96-c0bc09591133.png)

**System info (please complete the following information):**

-   OS: Ubuntu 20.10
-   Neovim version: NVIM v0.6.0

Checkhealth output (`:checkhealth nvim-lsp-installer`):
![image](https://user-images.githubusercontent.com/44520002/146349544-1f86f89a-3dd9-4a61-8d37-eb23b61ad6c6.png)

```
Paste `:checkhealth nvim-lsp-installer` output here
```
```

nvim-lsp-installer: require("nvim-lsp-installer.health").check()
========================================================================
## nvim-lsp-installer report
  - ERROR: **Composer**: not available
  - ERROR: **java**: not available
  - ERROR: **sh**: not available
  - INFO: **tar**: `tar (GNU tar) 1.34`
  - INFO: **gzip**: `gzip 1.10`
  - INFO: **wget**: `GNU Wget 1.21 built on linux-gnu.`
  - INFO: **python3**: `Python 3.8.8`
  - INFO: **PHP**: `PHP 7.4.16 (cli) (built: Oct 26 2021 16:46:20) ( NTS )`
  - INFO: **Ruby**: `ruby 2.7.2p137 (2020-10-01 revision 5445e04352) [x86_64-linux-gnu]`
  - INFO: **Go**: `go version go1.17.5 linux/amd64`
  - INFO: **bash**: `GNU bash, version 5.1.4(1)-release (x86_64-pc-linux-gnu)`
  - INFO: **curl**: `curl 7.71.1 (x86_64-conda_cos6-linux-gnu) libcurl/7.71.1 OpenSSL/1.1.1k zlib/1.2.11 libssh2/1.9.0`
  - INFO: **RubyGem**: `3.2.5`
  - INFO: **pip3**: `pip 21.0.1 from /home/phanlong2811/anaconda3/lib/python3.8/site-packages/pip (python 3.8)`
  - INFO: **node**: `v14.18.2`
  - INFO: **npm**: `6.14.15`


```
Please also include any other system information you think is relevant!

**Install log**
If applicable, include nvim-lsp-installer's logs. See `:help nvim-lsp-installer-debugging` for instructions on how to set the log level and access the logs.

```
Paste your logs here
```
        Submodule path 'third_party/rapidjson/thirdparty/gtest': checked out '0a439623f75c029912728d80cb7f1b8b48739ca4'
        -- The CXX compiler identification is GNU 10.3.0
        -- The C compiler identification is GNU 10.3.0
        -- Detecting CXX compiler ABI info
        -- Detecting CXX compiler ABI info - done
        -- Check for working CXX compiler: /usr/bin/c++ - skipped
        -- Detecting CXX compile features
        -- Detecting CXX compile features - done
        -- Detecting C compiler ABI info
        -- Detecting C compiler ABI info - done
        -- Check for working C compiler: /usr/bin/cc - skipped
        -- Detecting C compile features
        -- Detecting C compile features - done
        -- Found ZLIB: /usr/lib/x86_64-linux-gnu/libz.so (found version "1.2.11") 
        -- Could NOT find LibXml2 (missing: LIBXML2_LIBRARY LIBXML2_INCLUDE_DIR) 
        -- Looking for pthread.h
        -- Looking for pthread.h - found
        -- Performing Test CMAKE_HAVE_LIBC_PTHREAD
        -- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
        -- Check if compiler accepts -pthread
        -- Check if compiler accepts -pthread - yes
        -- Found Threads: TRUE  
        -- RapidJSON found. Headers: /usr/include
        -- Configuring done
        fatal: No names found, cannot describe anything.
        -- Generating done
        -- Build files have been written to: /home/phanlong2811/.local/share/nvim/lsp_servers/ccls/ccls/Release
        Scanning dependencies of target ccls
        [  2%] Building CXX object CMakeFiles/ccls.dir/third_party/siphash.cc.o
        [  4%] Building CXX object CMakeFiles/ccls.dir/src/clang_tu.cc.o
        [  6%] Building CXX object CMakeFiles/ccls.dir/src/config.cc.o
        [  9%] Building CXX object CMakeFiles/ccls.dir/src/filesystem.cc.o
        [ 11%] Building CXX object CMakeFiles/ccls.dir/src/fuzzy_match.cc.o
        [ 13%] Building CXX object CMakeFiles/ccls.dir/src/main.cc.o
        [ 16%] Building CXX object CMakeFiles/ccls.dir/src/include_complete.cc.o
        [ 18%] Building CXX object CMakeFiles/ccls.dir/src/indexer.cc.o
        [ 20%] Building CXX object CMakeFiles/ccls.dir/src/log.cc.o
        [ 23%] Building CXX object CMakeFiles/ccls.dir/src/lsp.cc.o
        [ 25%] Building CXX object CMakeFiles/ccls.dir/src/message_handler.cc.o
        [ 27%] Building CXX object CMakeFiles/ccls.dir/src/pipeline.cc.o
        [ 30%] Building CXX object CMakeFiles/ccls.dir/src/platform_posix.cc.o
        [ 32%] Building CXX object CMakeFiles/ccls.dir/src/platform_win.cc.o
        [ 34%] Building CXX object CMakeFiles/ccls.dir/src/position.cc.o
        [ 37%] Building CXX object CMakeFiles/ccls.dir/src/project.cc.o
        [ 39%] Building CXX object CMakeFiles/ccls.dir/src/query.cc.o
        In file included from /home/phanlong2811/.local/share/nvim/lsp_servers/ccls/ccls/src/indexer.hh:6,
                         from /home/phanlong2811/.local/share/nvim/lsp_servers/ccls/ccls/src/query.hh:6,
                         from /home/phanlong2811/.local/share/nvim/lsp_servers/ccls/ccls/src/query.cc:4:
        /home/phanlong2811/.local/share/nvim/lsp_servers/ccls/ccls/src/lsp.hh: In function ‘std::optional<ccls::SymbolInformation> ccls::getSymbolInfo(ccls::DB*, ccls::SymbolIdx, bool)’:
        /home/phanlong2811/.local/share/nvim/lsp_servers/ccls/ccls/src/lsp.hh:179:8: warning: ‘info.ccls::SymbolInformation::kind’ may be used uninitialized in this function [-Wmaybe-uninitialized]
          179 | struct SymbolInformation {
              |        ^~~~~~~~~~~~~~~~~
        [ 41%] Building CXX object CMakeFiles/ccls.dir/src/sema_manager.cc.o
        [ 44%] Building CXX object CMakeFiles/ccls.dir/src/serializer.cc.o
        [ 46%] Building CXX object CMakeFiles/ccls.dir/src/test.cc.o
        [ 48%] Building CXX object CMakeFiles/ccls.dir/src/utils.cc.o
        [ 51%] Building CXX object CMakeFiles/ccls.dir/src/working_files.cc.o
        [ 53%] Building CXX object CMakeFiles/ccls.dir/src/messages/ccls_call.cc.o
        [ 55%] Building CXX object CMakeFiles/ccls.dir/src/messages/ccls_info.cc.o
        [ 58%] Building CXX object CMakeFiles/ccls.dir/src/messages/ccls_inheritance.cc.o
        [ 60%] Building CXX object CMakeFiles/ccls.dir/src/messages/ccls_member.cc.o
        [ 62%] Building CXX object CMakeFiles/ccls.dir/src/messages/ccls_navigate.cc.o
        [ 65%] Building CXX object CMakeFiles/ccls.dir/src/messages/ccls_reload.cc.o
        [ 67%] Building CXX object CMakeFiles/ccls.dir/src/messages/ccls_vars.cc.o
        [ 69%] Building CXX object CMakeFiles/ccls.dir/src/messages/initialize.cc.o
        [ 72%] Building CXX object CMakeFiles/ccls.dir/src/messages/textDocument_code.cc.o
        [ 74%] Building CXX object CMakeFiles/ccls.dir/src/messages/textDocument_completion.cc.o
        [ 76%] Building CXX object CMakeFiles/ccls.dir/src/messages/textDocument_definition.cc.o
        [ 79%] Building CXX object CMakeFiles/ccls.dir/src/messages/textDocument_did.cc.o
        [ 81%] Building CXX object CMakeFiles/ccls.dir/src/messages/textDocument_foldingRange.cc.o
        [ 83%] Building CXX object CMakeFiles/ccls.dir/src/messages/textDocument_formatting.cc.o
        [ 86%] Building CXX object CMakeFiles/ccls.dir/src/messages/textDocument_document.cc.o
        [ 88%] Building CXX object CMakeFiles/ccls.dir/src/messages/textDocument_hover.cc.o
        [ 90%] Building CXX object CMakeFiles/ccls.dir/src/messages/textDocument_references.cc.o
        [ 93%] Building CXX object CMakeFiles/ccls.dir/src/messages/textDocument_rename.cc.o
        [ 95%] Building CXX object CMakeFiles/ccls.dir/src/messages/textDocument_signatureHelp.cc.o
        [ 97%] Building CXX object CMakeFiles/ccls.dir/src/messages/workspace.cc.o
        [100%] Linking CXX executable ccls
        /usr/bin/ld: cannot find -lexecinfo
        collect2: error: ld returned 1 exit status
        gmake[2]: *** [CMakeFiles/ccls.dir/build.make:761: ccls] Error 1
        gmake[1]: *** [CMakeFiles/Makefile2:207: CMakeFiles/ccls.dir/all] Error 2
        gmake: *** [Makefile:149: all] Error 2

--- a/lua/nvim-lsp-installer/installers/context.lua
+++ b/lua/nvim-lsp-installer/installers/context.lua
@@ -169,7 +169,7 @@ end
 ---Creates an installer that moves the current installation directory to the server's root directory.
 function M.promote_install_dir()
     ---@type ServerInstallerFunction
-    return function(server, callback, context)
+    return vim.schedule_wrap(function(server, callback, context)
         if server:promote_install_dir(context.install_dir) then
             context.install_dir = server.root_dir
             callback(true)
@@ -179,7 +179,7 @@ function M.promote_install_dir()
             )
             callback(false)
         end
-    end
+    end)
 end
 
 ---Access the context ojbect to create a new installer.
@@ -223,6 +223,78 @@ function M.set_working_dir(rel_path)
         context.install_dir = new_dir
         callback(true)
     end)
+end
+
+function M.use_os_distribution()
+    ---Parses the provided contents of an /etc/\*-release file and identifies the Linux distribution.
+    ---@param contents string @The contents of a /etc/\*-release file.
+    ---@return table<string, any>
+    local function parse_linux_dist(contents)
+        local lines = vim.split(contents, "\n")
+
+        local entries = {}
+
+        for i = 1, #lines do
+            local line = lines[i]
+            local index = line:find "="
+            if index then
+                local key = line:sub(1, index - 1)
+                local value = line:sub(index + 1)
+                entries[key] = value
+            end
+        end
+
+        if entries.ID == "ubuntu" then
+            -- Parses the Ubuntu OS VERSION_ID into their version components, e.g. "18.04" -> {major=18, minor=04}
+            local version_id = entries.VERSION_ID:gsub([["]], "")
+            local version_parts = vim.split(version_id, "%.")
+            local major = tonumber(version_parts[1])
+            local minor = tonumber(version_parts[2])
+
+            return {
+                id = "ubuntu",
+                version_id = version_id,
+                version = { major = major, minor = minor },
+            }
+        else
+            return {
+                id = "linux-generic",
+            }
+        end
+    end
+
+    return installers.when {
+        ---@type ServerInstallerFunction
+        linux = function(_, callback, ctx)
+            local stdio = process.in_memory_sink()
+            process.spawn("bash", {
+                args = { "-c", "cat /etc/*-release" },
+                stdio_sink = stdio.sink,
+            }, function(success)
+                if success then
+                    ctx.os_distribution = parse_linux_dist(table.concat(stdio.buffers.stdout, ""))
+                    callback(true)
+                else
+                    ctx.os_distribution = {
+                        id = "linux-generic",
+                    }
+                    callback(true)
+                end
+            end)
+        end,
+        mac = function(_, callback, ctx)
+            ctx.os_distribution = {
+                id = "macOS",
+            }
+            callback(true)
+        end,
+        win = function(_, callback, ctx)
+            ctx.os_distribution = {
+                id = "windows",
+            }
+            callback(true)
+        end,
+    }
 end
 
 return M

--- a/lua/nvim-lsp-installer/installers/init.lua
+++ b/lua/nvim-lsp-installer/installers/init.lua
@@ -1,9 +1,6 @@
 local platform = require "nvim-lsp-installer.platform"
 local log = require "nvim-lsp-installer.log"
 local Data = require "nvim-lsp-installer.data"
-local process = require "nvim-lsp-installer.process"
-local fs = require "nvim-lsp-installer.fs"
-local path = require "nvim-lsp-installer.path"
 
 local M = {}
 
@@ -23,6 +20,7 @@ end
 ---@field requested_server_version string|nil @The version requested by the user.
 ---@field stdio_sink StdioSink
 ---@field github_release_file string|nil @Only available if context.use_github_release_file has been called.
+---@field os_distribution table<string, any> @Only available if context.use_os_distribution has been called.
 ---@field install_dir string
 
 ---@alias ServerInstallerFunction fun(server: Server, callback: ServerInstallCallback, context: ServerInstallContext)

--- a/lua/nvim-lsp-installer/servers/ccls/init.lua
+++ b/lua/nvim-lsp-installer/servers/ccls/init.lua
@@ -14,15 +14,32 @@ return function(name, root_dir)
 
     do
         ---@param version string
+        ---@param os_distribution table<string, string>|nil
         ---@return string|nil
-        local function get_archive_name(version)
+        local function get_archive_name(version, os_distribution)
             local name_template = coalesce(
                 when(platform.is_mac, "clang+llvm-%s-x86_64-apple-darwin"),
                 when(
                     platform.is_linux,
                     coalesce(
-                        when(platform.arch == "x64", "clang+llvm-%s-amd64-unknown-freebsd13"),
-                        when(platform.arch == "arm64", "clang+llvm-%s-aarch64-linux-gnu")
+                        when(
+                            platform.arch == "x64",
+                            coalesce(
+                                when(
+                                    os_distribution.id == "ubuntu" and os_distribution.version.major >= 20,
+                                    "clang+llvm-%s-x86_64-linux-gnu-ubuntu-20.04"
+                                ),
+                                when(
+                                    os_distribution.id == "ubuntu" and os_distribution.version.major >= 16,
+                                    "clang+llvm-%s-x86_64-linux-gnu-ubuntu-16.04"
+                                ),
+                                -- the Ubuntu dist is allegedly the most suitable cross-platform one, so we default to it
+                                "clang+llvm-%s-x86_64-linux-gnu-ubuntu-16.04"
+                            )
+                        ),
+                        when(platform.arch == "arm64", "clang+llvm-%s-aarch64-linux-gnu"),
+                        when(platform.arch == "armv7", "clang+llvm-%s-armv7a-linux-gnueabihf"),
+                        when(platform.arch == "x86", "clang+llvm-%s-i386-unknown-freebsd13")
                     )
                 )
             )
@@ -31,7 +48,8 @@ return function(name, root_dir)
 
         ---@param version string
         local function normalize_version(version)
-            return version:gsub("^llvmorg%-", "")
+            local s = version:gsub("^llvmorg%-", "")
+            return s
         end
 
         llvm_installer = installers.branch_context {
@@ -39,15 +57,24 @@ return function(name, root_dir)
                 -- We unset the requested version for llvm because it's not the primary target - the user's requested version should only apply to ccls.
                 ctx.requested_server_version = nil
             end),
-            context.use_github_release_file("llvm/llvm-project", function(version)
-                -- Strip the "llvmorg-" prefix from tags (llvm releases tags like llvmorg-13.0.0)
-                local archive_name = get_archive_name(normalize_version(version))
-                return archive_name and ("%s.tar.xz"):format(archive_name)
+            context.capture(function(ctx)
+                return context.use_github_release_file("llvm/llvm-project", function(version)
+                    -- Strip the "llvmorg-" prefix from tags (llvm releases tags like llvmorg-13.0.0)
+                    local archive_name = get_archive_name(normalize_version(version), ctx.os_distribution)
+                    return archive_name and ("%s.tar.xz"):format(archive_name)
+                end)
             end),
             context.capture(function(ctx)
                 return installers.pipe {
                     std.untarxz_remote(ctx.github_release_file),
-                    std.rename(get_archive_name(normalize_version(ctx.requested_server_version)), "llvm"),
+                    std.rename(
+                        get_archive_name(normalize_version(ctx.requested_server_version), ctx.os_distribution),
+                        "llvm"
+                    ),
+                    std.rename(
+                        path.concat { "llvm", "lib", "clang", normalize_version(ctx.requested_server_version) },
+                        "clang-resource"
+                    ),
                 }
             end),
         }
@@ -56,25 +83,30 @@ return function(name, root_dir)
     local ccls_installer = installers.branch_context {
         context.set(function(ctx)
             ctx.llvm_install_dir = path.concat { ctx.install_dir, "llvm" }
+            ctx.clang_resource_dir = path.concat { ctx.install_dir, "clang-resource" }
         end),
-        context.set_working_dir "ccls",
-        std.git_clone "https://github.com/MaskRay/ccls",
-        std.git_submodule_update(),
-        function(_, callback, ctx)
-            local c = process.chain {
-                cwd = ctx.install_dir,
-                stdio_sink = ctx.stdio_sink,
-            }
+        installers.branch_context {
+            context.set_working_dir "ccls-git",
+            std.git_clone "https://github.com/MaskRay/ccls",
+            std.git_submodule_update(),
+            function(_, callback, ctx)
+                local c = process.chain {
+                    cwd = ctx.install_dir,
+                    stdio_sink = ctx.stdio_sink,
+                }
 
-            c.run("cmake", {
-                "-H.",
-                "-BRelease",
-                "-DCMAKE_BUILD_TYPE=Release",
-                ("-DCMAKE_PREFIX_PATH=%s"):format(path.concat { ctx.llvm_install_dir, "lib", "cmake" }),
-            })
-            c.run("cmake", { "--build", "Release" })
-            c.spawn(callback)
-        end,
+                c.run("cmake", {
+                    "-H.",
+                    "-BRelease",
+                    "-DCMAKE_BUILD_TYPE=Release",
+                    ("-DCMAKE_PREFIX_PATH=%s"):format(ctx.llvm_install_dir),
+                    ("-DCLANG_RESOURCE_DIR=%s"):format(ctx.clang_resource_dir),
+                })
+                c.run("cmake", { "--build", "Release" })
+                c.spawn(callback)
+            end,
+        },
+        std.rename(path.concat { "ccls-git", "Release", "ccls" }, "ccls"),
     }
 
     return server.Server:new {
@@ -84,13 +116,16 @@ return function(name, root_dir)
         languages = { "c", "c++", "objective-c" },
         installer = installers.when {
             unix = {
+                context.use_os_distribution(),
                 context.promote_install_dir(), -- ccls hardcodes the path to llvm at compile time, so we need to compile everything in the final directory
                 llvm_installer,
                 ccls_installer,
+                std.rmrf "llvm",
+                std.rmrf "ccls-git",
             },
         },
         default_options = {
-            cmd = { path.concat { root_dir, "ccls", "Release", "ccls" } },
+            cmd = { path.concat { root_dir, "ccls" } },
         },
     }
 end

